server/server.go

```diff
@@ -4138,8 +4138,10 @@ func (s *Server) lameDuckMode() {
 	if s.websocket.server != nil {
 		expected++
 		s.websocket.server.Close()
+		s.websocket.mu.Lock()
```

Comment:
Still think this would be better.

```go
s.websocket.mu.Lock()
if s.websocket.server != nil {
		expected++
		s.websocket.server.Close()
		s.websocket.server = nil
		s.websocket.listener = nil
	}
s.websocket.mu.Unlock()
```

Reply:
I am concerned because `s.websocket.server.Close()` internally waits on listener to be done and I did not want to risk to have a deadlock somewhere.

Reply:
Capture server in local variable under lock and call Close() after you release lock?

Reply:
Actually, I had another look and it is perfectly fine to move the lock, so just pushed a new commit.

Reply:
I could even make it a small function instead of having the same code in shutdown() and lameDuckMode(). Let me do that.